### PR TITLE
Add support for Exasol

### PR DESCRIPTION
**Pull request by kevang**
### CATEGORY

Choose one

- [ ] Bug Fix
- [✓] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Exasol describes itself as a "high-performance, cloud-first analytics database", which could be a good addition to Superset's arsenal of databases. sqlalchemy_exasol (https://github.com/blue-yonder/sqlalchemy_exasol), which uses ODBC, is the sqlalchemy dialect that we used. A change in the SupersetDataFrame class is necessary to get access to data via the SQL Lab.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [✓] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/db_engine_specs/exasol.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+from superset.db_engine_specs.base import BaseEngineSpec
+
+
+class ExasolEngineSpec(BaseEngineSpec):
+    """Engine spec for Exasol"""
+
+    engine = "exa"
+    max_column_name_length = 128
+    odbc_row_results = True
```
Comment (villebro) [inline superset/db_engine_specs/exasol.py]: When adding new properties, you should add a default value (`False`) to `BaseEngineSpec` in `base.py`.

File: superset/dataframe.py [hunk 2]
```diff
@@ -119,6 +119,8 @@ def __init__(self, data, cursor_description, db_engine_spec):
             }
             self.df = pd.DataFrame(data, columns=column_names)
         else:
+            if getattr(db_engine_spec, "odbc_row_results", None):
+                data = [[value for value in row] for row in data]
```
Comment (villebro) [inline superset/dataframe.py]: Can you explain why this is needed? Also, if the attribute is added to `BaseEngineSpec`, you should not need to test for it first (see below).

File: superset/db_engine_specs/exasol.py (hunk 1 above)
Comment (villebro) [reply superset/db_engine_specs/exasol.py]: At a minimum, all engine specs should define at least a minimum set of time grains. Take a look at e.g. `sqlite.py`.

**Review by villebro — CHANGES_REQUESTED** [review]
Left some first pass comments.

File: superset/db_engine_specs/exasol.py (hunk 1 above)
Comment (kevang) [reply superset/db_engine_specs/exasol.py]: Thanks, I'll check and update accordingly.

File: superset/dataframe.py (hunk 2 above)
Comment (villebro) [reply superset/dataframe.py]: The problem with fixing this behaviour per engine is that several engines support both odbc and non-odbc drivers (e.g. mssql referenced in the test). Does the default behaviour not work for exa? As pyodbc has been supported for quite some time on other engines, it feels strange that this should need to be handled differently in this case.

Comment (kevang) [reply superset/dataframe.py]: The default behaviour fails during `pd.Dataframe` instantiation with messages similar to `Shape of passed values is (1, 1), indices imply (1, 50)`, or when I `SELECT` only one column (where the two cases converge and the dataframe can be created) it fails at a later stage with `Unserializable [<class 'pyodbc.Row'>]`

While debugging I also tried to leave the data list as is and pass column names like `[["col1"], ["col2"], ["col3"]]` as per the test, but this was not successful either. 

> As pyodbc has been supported for quite some time on other engines, it feels strange that this should need to be handled differently in this case.

🤷‍♂

Comment (kevang) [reply superset/dataframe.py]: Would it be preferable to add a try/expect block instead?

File: docs/installation.rst [hunk 3]
```diff
@@ -659,6 +661,25 @@ it in the ``extra`` parameter::
...
+
+Example config (odbcinst.ini can be left empty) ::
+
+    $ cat $/.../path/to/odbc.ini
+    [EXAODBC]
+    DRIVER = /.../path/to/driver/EXASOL_driver.so
+    EXAHOST = host:8563
+    EXASCHEMA = main
+
+See `SQLAlchemy for Exasol <https://https://github.com/blue-yonder/sqlalchemy_exasol>`_.
```
Comment (villebro) [inline docs/installation.rst]: There is one `https://` too many here

**Comment by villebro** [comment]
I tried starting the exasol dev container from here: https://github.com/exasol/docker-db, installed the exasol 6.2.1 driver (MacOS) and the blue-yonder sqlalchemy connector, but was not able to connect Superset to it. Any idea what the `odbc.ini` file should look like on MacOS? Not sure where the `.so` file goes.. Anyway, this is the error I'm getting on Superset: 
image

**Comment by kevang** [comment]
Many thanks for taking the time to replicate the issue. 

I'm also on Mac and set up superset with http://superset.incubator.apache.org/installation.html#start-with-docker. Inside the container I created `odbc.ini` with the `config_odbc` tool (which, together with the driver can be downloaded from https://docs.exasol.com/connect_exasol/drivers/odbc/odbc_linux.htm). After running the above script, `odbcinst -j` should list `USER DATA SOURCES..: /root/.odbc.ini`. As a final step the contents of `odbc.ini` should be manually changed (after variables are changed accordingly) to

```
root@307bbaccb9d6:/home/superset# cat /root/.odbc.ini
[EXAODBC]
DRIVER = /root/.driver/EXASOL_ODBC-6.1.5/lib/linux/x86_64/libexaodbc-uo2214lv2.so
EXAHOST = host.docker.internal:8563
EXASCHEMA = schema
```

 Then in Python's REPL I can query with:
```
>>> from sqlalchemy import create_engine
>>> e = create_engine("exa+pyodbc://user:password@EXAODBC")
>>> e.connect()
>>> e.execute("select * from table LIMIT 1").fetchall()
```

File: superset/dataframe.py (hunk 2 above)
Comment (villebro) [reply superset/dataframe.py]: That still feels hacky, so would prefer not to go down that route. My intuition tells me there might be a bug in the SqlAlchemy connector. Ideally we would like to end up with a solution that works identically for all odbc sources.

File: superset/db_engine_specs/exasol.py (hunk 1 above)
Comment (villebro) [reply superset/db_engine_specs/exasol.py]: Btw, If memory serves me right Exasol is inherently an Oracle dialect, so you might be best off extending the Oracle spec.

Comment (kevang) [reply superset/db_engine_specs/exasol.py]: Afaik this is not the case, Exasol borrows heavily from postgres (e.g. `DATE_TRUNC` which I used to implement the time grain functions is the same in both), but is not a clear descendant in the way that for example Redshift is.

**Comment by kevang** [comment]
I added the `_time_grain_functions` and implemented a try/except solution which at least is driver independent. I understand that this less than ideal, so happy to hear your thoughts.

File: superset/dataframe.py (hunk 2 above)
Comment (mistercrunch) [reply superset/dataframe.py]: If something special is needed for Exasol/odbc it should be in the engine specific module. One way would be to override `fetch_data` for Exasol.
https://github.com/apache/incubator-superset/blob/master/superset/db_engine_specs/base.py#L225

**Review by mistercrunch — CHANGES_REQUESTED** [review]

File: superset/db_engine_specs/exasol.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        "P1D": "DATE_TRUNC('day', {col})",
+        "P1W": "DATE_TRUNC('week', {col})",
+        "P1M": "DATE_TRUNC('month', {col})",
+        "P0.25Y": "DATE_TRUNC('quarter', {col})",
+        "P1Y": "DATE_TRUNC('year', {col})",
+    }
+
+    @classmethod
+    def fetch_data(cls, cursor, limit: int) -> List[Tuple]:
+        data = super(ExasolEngineSpec, cls).fetch_data(cursor, limit)
```
Comment (villebro) [inline superset/db_engine_specs/exasol.py]: Nit: this is equivalent to `super().fetch_data(cursor, limit)`.

**Review by villebro — CHANGES_REQUESTED** [review]
This looks like a good solution; other than the nit LGTM!

File: superset/db_engine_specs/exasol.py (hunk 4 above)
Comment (kevang) [reply superset/db_engine_specs/exasol.py]: Yes ofc, python 2 habits die hard :D Thanks, updated it.

**Review by villebro — APPROVED** [review]

**Comment by mistercrunch** [comment]
yeah

**Pull request merged**